Atomic operations detections fails on OpenWRT/LEDE

https://github.com/firehol/netdata/commit/db787ff47f80740ae266e0e347a6b49b3a395c35

After this commit netdata fails to compile using the LEDE-project toolchain.

`global_statistics.o: In function `finished_web_request_statistics':
global_statistics.c:(.text+0x90): undefined reference to `__atomic_compare_exchange_8'
global_statistics.c:(.text+0xc8): undefined reference to `__atomic_fetch_add_8'
global_statistics.c:(.text+0xe0): undefined reference to `__atomic_fetch_add_8'
global_statistics.c:(.text+0xf8): undefined reference to `__atomic_fetch_add_8'
global_statistics.c:(.text+0x110): undefined reference to `__atomic_fetch_add_8'
global_statistics.c:(.text+0x128): undefined reference to `__atomic_fetch_add_8'
global_statistics.o:global_statistics.c:(.text+0x140): more undefined references to `__atomic_fetch_add_8' follow
global_statistics.o: In function `global_statistics_copy':
global_statistics.c:(.text+0x3a4): undefined reference to `__atomic_compare_exchange_8'
collect2: error: ld returned 1 exit status
Makefile:575: recipe for target 'netdata' failed
make[6]: *** [netdata] Error 1
`

--- a/configure.ac
+++ b/configure.ac
@@ -82,6 +82,7 @@ AC_TYPE_UINT8_T
 AC_TYPE_UINT16_T
 AC_TYPE_UINT32_T
 AC_C_INLINE
+AC_FUNC_STRERROR_R
 
 # https://lists.gnu.org/archive/html/autoconf-commit/2012-12/msg00004.html
 # AC_C__GENERIC
@@ -112,7 +113,39 @@ fi
 ])# AC_C__GENERIC
 
 AC_C__GENERIC
-AC_FUNC_STRERROR_R
+
+# AC_C___ATOMIC
+# -------------
+# Define HAVE_C___ATOMIC if __atomic works.
+AN_IDENTIFIER([__atomic], [AC_C___ATOMIC])
+AC_DEFUN([AC_C___ATOMIC],
+[AC_CACHE_CHECK([for __atomic], ac_cv_c___atomic,
+[AC_COMPILE_IFELSE(
+   [AC_LANG_SOURCE(
+      [[int
+        main (int argc, char **argv)
+        {
+          volatile unsigned long ul1 = 1, ul2 = 0, ul3 = 2;
+          __atomic_compare_exchange(&ul1, &ul2, &ul3, 1, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+          __atomic_fetch_add(&ul1, 1, __ATOMIC_SEQ_CST);
+          __atomic_fetch_sub(&ul3, 1, __ATOMIC_SEQ_CST);
+          volatile unsigned long long ull1 = 1, ull2 = 0, ull3 = 2;
+          __atomic_compare_exchange(&ull1, &ull2, &ull3, 1, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
+          __atomic_fetch_add(&ull1, 1, __ATOMIC_SEQ_CST);
+          __atomic_fetch_sub(&ull3, 1, __ATOMIC_SEQ_CST);
+          return 0;
+        }
+      ]])],
+   [ac_cv_c___atomic=yes],
+   [ac_cv_c___atomic=no])])
+if test $ac_cv_c___atomic = yes; then
+  AC_DEFINE([HAVE_C___ATOMIC], 1,
+           [Define to 1 if __atomic operations work.])
+fi
+])# AC_C___ATOMIC
+
+AC_C___ATOMIC
+
 
 AC_ARG_VAR([MATH_CFLAGS], [C compiler flags for math])
 AC_ARG_VAR([MATH_LIBS], [linker flags for math])

--- a/configure.ac
+++ b/configure.ac
@@ -83,6 +83,37 @@ AC_TYPE_UINT16_T
 AC_TYPE_UINT32_T
 AC_C_INLINE
 
+# https://lists.gnu.org/archive/html/autoconf-commit/2012-12/msg00004.html
+# AC_C__GENERIC
+# -------------
+# Define HAVE_C__GENERIC if _Generic works, a la C11.
+AN_IDENTIFIER([_Generic], [AC_C__GENERIC])
+AC_DEFUN([AC_C__GENERIC],
+[AC_CACHE_CHECK([for _Generic], ac_cv_c__Generic,
+[AC_COMPILE_IFELSE(
+   [AC_LANG_SOURCE(
+      [[int
+        main (int argc, char **argv)
+        {
+          int a = _Generic (argc, int: argc = 1);
+          int *b = &_Generic (argc, default: argc);
+          char ***c = _Generic (argv, int: argc, default: argv ? &argv : 0);
+          _Generic (1 ? 0 : b, int: a, default: b) = &argc;
+          _Generic (a = 1, default: a) = 3;
+          return a + !b + !c;
+        }
+      ]])],
+   [ac_cv_c__Generic=yes],
+   [ac_cv_c__Generic=no])])
+if test $ac_cv_c__Generic = yes; then
+  AC_DEFINE([HAVE_C__GENERIC], 1,
+           [Define to 1 if C11-style _Generic works.])
+fi
+])# AC_C__GENERIC
+
+AC_C__GENERIC
+AC_FUNC_STRERROR_R
+
 AC_ARG_VAR([MATH_CFLAGS], [C compiler flags for math])
 AC_ARG_VAR([MATH_LIBS], [linker flags for math])
 if test -z "${MATH_LIBS}"; then

--- a/plugins.d/loopsleepms.sh.inc
+++ b/plugins.d/loopsleepms.sh.inc
@@ -13,6 +13,7 @@ fi
 now_ms=
 LOOPSLEEPMS_HIGHRES=1
 test "$($LOOPSLEEP_DATE +%N)" = "%N" && LOOPSLEEPMS_HIGHRES=0
+test -z "$($LOOPSLEEP_DATE +%N)"     && LOOPSLEEPMS_HIGHRES=0
 current_time_ms_from_date() {
     if [ $LOOPSLEEPMS_HIGHRES -eq 0 ]
     then

--- a/src/common.h
+++ b/src/common.h
@@ -65,9 +65,11 @@
 #endif
 
 #ifdef __GNUC__
-#define GCC_VERSION (__GNUC__ * 10000 \
-                               + __GNUC_MINOR__ * 100 \
-                               + __GNUC_PATCHLEVEL__)
+#define GCC_VERSION (__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__)
+
+#if GCC_VERSION < 40900
+#define WITHOUT_C11_GENERIC 1
+#endif
 
 #if __x86_64__ || __ppc64__
 #define ENVIRONMENT64

--- a/src/common.h
+++ b/src/common.h
@@ -60,10 +60,6 @@
 #include <zlib.h>
 #endif
 
-#ifndef __ATOMIC_SEQ_CST
-#define NETDATA_NO_ATOMIC_INSTRUCTIONS 1
-#endif
-
 #ifdef __GNUC__
 #define GCC_VERSION (__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__)
 

--- a/src/common.h
+++ b/src/common.h
@@ -67,10 +67,6 @@
 #ifdef __GNUC__
 #define GCC_VERSION (__GNUC__ * 10000 + __GNUC_MINOR__ * 100 + __GNUC_PATCHLEVEL__)
 
-#if GCC_VERSION < 40900
-#define WITHOUT_C11_GENERIC 1
-#endif
-
 #if __x86_64__ || __ppc64__
 #define ENVIRONMENT64
 #else

--- a/src/global_statistics.c
+++ b/src/global_statistics.c
@@ -25,7 +25,7 @@ void finished_web_request_statistics(uint64_t dt,
                                      uint64_t bytes_sent,
                                      uint64_t content_size,
                                      uint64_t compressed_content_size) {
-#ifndef NETDATA_NO_ATOMIC_INSTRUCTIONS
+#if defined(HAVE_C___ATOMIC) && !defined(NETDATA_NO_ATOMIC_INSTRUCTIONS)
     uint64_t old_web_usec_max = global_statistics.web_usec_max;
     while(dt > old_web_usec_max)
         __atomic_compare_exchange(&global_statistics.web_usec_max, &old_web_usec_max, &dt, 1, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST);
@@ -57,7 +57,7 @@ void finished_web_request_statistics(uint64_t dt,
 }
 
 void web_client_connected(void) {
-#ifndef NETDATA_NO_ATOMIC_INSTRUCTIONS
+#if defined(HAVE_C___ATOMIC) && !defined(NETDATA_NO_ATOMIC_INSTRUCTIONS)
     __atomic_fetch_add(&global_statistics.connected_clients, 1, __ATOMIC_SEQ_CST);
 #else
     if (web_server_mode == WEB_SERVER_MODE_MULTI_THREADED)
@@ -71,7 +71,7 @@ void web_client_connected(void) {
 }
 
 void web_client_disconnected(void) {
-#ifndef NETDATA_NO_ATOMIC_INSTRUCTIONS
+#if defined(HAVE_C___ATOMIC) && !defined(NETDATA_NO_ATOMIC_INSTRUCTIONS)
     __atomic_fetch_sub(&global_statistics.connected_clients, 1, __ATOMIC_SEQ_CST);
 #else
     if (web_server_mode == WEB_SERVER_MODE_MULTI_THREADED)
@@ -86,7 +86,7 @@ void web_client_disconnected(void) {
 
 
 inline void global_statistics_copy(struct global_statistics *gs, uint8_t options) {
-#ifndef NETDATA_NO_ATOMIC_INSTRUCTIONS
+#if defined(HAVE_C___ATOMIC) && !defined(NETDATA_NO_ATOMIC_INSTRUCTIONS)
     gs->connected_clients       = __atomic_fetch_add(&global_statistics.connected_clients, 0, __ATOMIC_SEQ_CST);
     gs->web_requests            = __atomic_fetch_add(&global_statistics.web_requests, 0, __ATOMIC_SEQ_CST);
     gs->web_usec                = __atomic_fetch_add(&global_statistics.web_usec, 0, __ATOMIC_SEQ_CST);

--- a/src/health.c
+++ b/src/health.c
@@ -2,12 +2,96 @@
 
 #define RRDVAR_MAX_LENGTH 1024
 
-static const char *health_default_exec = PLUGINS_DIR "/alarm-notify.sh";
-static const char *health_default_recipient = "root";
+struct health_options {
+    const char *health_default_exec;
+    const char *health_default_recipient;
+    const char *log_filename;
+    FILE *log_fp;
+};
+
+static struct health_options health = {
+    .health_default_exec = PLUGINS_DIR "/alarm-notify.sh",
+    .health_default_recipient = "root",
+    .log_filename = VARLIB_DIR "/health/alarm_log.db",
+    .log_fp = NULL
+};
+
 int health_enabled = 1;
 
 // ----------------------------------------------------------------------------
-// Health Alarms Log Management
+// health alarm log load/save
+// no need for locking - only one thread is reading / writing the alarms log
+
+static inline int health_alarm_log_open(void) {
+    if(health.log_fp)
+        fclose(health.log_fp);
+
+    health.log_fp = fopen(health.log_filename, "a");
+
+    if(health.log_fp) {
+        if (setvbuf(health.log_fp, NULL, _IOLBF, 0) != 0)
+            error("Cannot set line buffering on health log file.");
+        return 0;
+    }
+
+    error("Cannot open health log file '%s'. Health data will be lost in case of netdata or server crash.", health.log_filename);
+    return -1;
+}
+
+static inline void health_alarm_log_close(void) {
+    if(health.log_fp) {
+        fclose(health.log_fp);
+        health.log_fp = NULL;
+    }
+}
+
+static inline void health_log_recreate(void) {
+    if(health.log_fp != NULL) {
+        health_alarm_log_close();
+
+        // open it with truncate
+        health.log_fp = fopen(health.log_filename, "w");
+        if(health.log_fp) fclose(health.log_fp);
+        else error("Cannot truncate health log '%s'", health.log_filename);
+
+        health.log_fp = NULL;
+
+        health_alarm_log_open();
+    }
+}
+
+static inline void health_alarm_log_save(RRDHOST *host, ALARM_ENTRY *ae) {
+    (void)host;
+    (void)ae;
+    
+/*    if(likely(health.log_fp)) {
+        if(unlikely(fprintf(health.log_fp, "A\t%s\t%08x\t%08x\t%08x\t%08x\t%08x\t%08x\t%s\t%s\t%s\t%s\t%s\t%08x\n",
+            host->hostname,
+            ae->unique_id,
+            ae->alarm_id,
+            ae->alarm_event_id,
+            (uint32_t)ae->when,
+            (uint32_t)ae->duration,
+            (uint32_t)ae->non_clear_duration,
+            (uint32_t)ae->exec_run_timestamp,
+            ae->name,
+            ae->chart,
+            ae->family,
+            ae->exec,
+            ae->recipient
+            ) < 0))
+            error("Health: failed to save alarm log entry. Health data may be lost in case of abnormal restart.");
+    }
+*/
+}
+
+static inline void health_alarm_log_load(RRDHOST *host) {
+    (void)host;
+
+}
+
+// ----------------------------------------------------------------------------
+// health alarm log management
 
 static inline void health_alarm_log(RRDHOST *host,
                 uint32_t alarm_id, uint32_t alarm_event_id,
@@ -75,6 +159,8 @@ static inline void health_alarm_log(RRDHOST *host,
                 if((t->new_status == RRDCALC_STATUS_WARNING || t->new_status == RRDCALC_STATUS_CRITICAL) &&
                    (t->old_status == RRDCALC_STATUS_WARNING || t->old_status == RRDCALC_STATUS_CRITICAL))
                     ae->non_clear_duration += t->non_clear_duration;
+
+                health_alarm_log_save(host, t);
             }
             else {
                 // no need to continue
@@ -83,6 +169,8 @@ static inline void health_alarm_log(RRDHOST *host,
         }
     }
     pthread_rwlock_unlock(&host->health_log.alarm_log_rwlock);
+
+    health_alarm_log_save(host, ae);
 }
 
 // ----------------------------------------------------------------------------
@@ -1709,12 +1797,14 @@ void health_init(void) {
         return;
     }
 
+    health_alarm_log_load(&localhost);
+
     char *path = health_config_dir();
 
     {
         char buffer[FILENAME_MAX + 1];
         snprintfz(buffer, FILENAME_MAX, "%s/alarm-notify.sh", config_get("global", "plugins directory", PLUGINS_DIR));
-        health_default_exec = config_get("health", "script to execute on alarm", buffer);
+        health.health_default_exec = config_get("health", "script to execute on alarm", buffer);
     }
 
     long n = config_get_number("health", "in memory max health log entries", (long)localhost.health_log.max);
@@ -1776,8 +1866,8 @@ static inline void health_alarm_entry2json_nolock(BUFFER *wb, ALARM_ENTRY *ae, R
                    (ae->notifications & HEALTH_ENTRY_NOTIFICATIONS_UPDATED)?"true":"false",
                    (unsigned long)ae->exec_run_timestamp,
                    (ae->notifications & HEALTH_ENTRY_NOTIFICATIONS_EXEC_FAILED)?"true":"false",
-                   ae->exec?ae->exec:health_default_exec,
-                   ae->recipient?ae->recipient:health_default_recipient,
+                   ae->exec?ae->exec:health.health_default_exec,
+                   ae->recipient?ae->recipient:health.health_default_recipient,
                    ae->exec_code,
                    ae->source,
                    ae->units?ae->units:"",
@@ -1852,8 +1942,8 @@ static inline void health_rrdcalc2json_nolock(BUFFER *wb, RRDCALC *rc) {
             , rc->chart
             , (rc->rrdset && rc->rrdset->family)?rc->rrdset->family:""
             , (rc->rrdset)?"true":"false"
-            , rc->exec?rc->exec:health_default_exec
-            , rc->recipient?rc->recipient:health_default_recipient
+            , rc->exec?rc->exec:health.health_default_exec
+            , rc->recipient?rc->recipient:health.health_default_recipient
             , rc->source
             , rc->units?rc->units:""
             , rc->info?rc->info:""
@@ -2016,7 +2106,6 @@ void health_reload(void) {
     }
 }
 
-
 // ----------------------------------------------------------------------------
 // health main thread and friends
 
@@ -2039,23 +2128,23 @@ static inline void health_alarm_execute(RRDHOST *host, ALARM_ENTRY *ae) {
     if(t && t->new_status == ae->new_status) {
         // don't send the same notification again
         info("Health not sending again notification for alarm '%s.%s' status %s", ae->chart, ae->name, rrdcalc_status2string(ae->new_status));
-        return;
+        goto done;
     }
 
     if((ae->old_status == RRDCALC_STATUS_UNDEFINED && ae->new_status == RRDCALC_STATUS_UNINITIALIZED)
         || (ae->old_status == RRDCALC_STATUS_UNINITIALIZED && ae->new_status == RRDCALC_STATUS_CLEAR)) {
         info("Health not sending notification for first initialization of alarm '%s.%s' status %s", ae->chart, ae->name, rrdcalc_status2string(ae->new_status));
-        return;
+        goto done;
     }
 
     char buffer[FILENAME_MAX + 1];
     pid_t command_pid;
 
     const char *exec = ae->exec;
-    if(!exec) exec = health_default_exec;
+    if(!exec) exec = health.health_default_exec;
 
     const char *recipient = ae->recipient;
-    if(!recipient) recipient = health_default_recipient;
+    if(!recipient) recipient = health.health_default_recipient;
 
     snprintfz(buffer, FILENAME_MAX, "exec %s '%s' '%s' '%u' '%u' '%u' '%lu' '%s' '%s' '%s' '%s' '%s' '%0.0Lf' '%0.0Lf' '%s' '%u' '%u' '%s' '%s'",
               exec,
@@ -2086,7 +2175,7 @@ static inline void health_alarm_execute(RRDHOST *host, ALARM_ENTRY *ae) {
     FILE *fp = mypopen(buffer, &command_pid);
     if(!fp) {
         error("HEALTH: Cannot popen(\"%s\", \"r\").", buffer);
-        return;
+        goto done;
     }
     debug(D_HEALTH, "HEALTH reading from command");
     char *s = fgets(buffer, FILENAME_MAX, fp);
@@ -2096,6 +2185,10 @@ static inline void health_alarm_execute(RRDHOST *host, ALARM_ENTRY *ae) {
 
     if(ae->exec_code != 0)
         ae->notifications |= HEALTH_ENTRY_NOTIFICATIONS_EXEC_FAILED;
+
+done:
+    health_alarm_log_save(host, ae);
+    return;
 }
 
 static inline void health_process_notifications(RRDHOST *host, ALARM_ENTRY *ae) {

--- a/src/log.c
+++ b/src/log.c
@@ -266,17 +266,13 @@ void info_int( const char *file, const char *function, const unsigned long line,
 // ----------------------------------------------------------------------------
 // error log
 
-#ifdef WITHOUT_C11_GENERIC
-
-#ifdef STRERROR_R_POSIX
-// POSIX version of strerror_r
-static const char *strerror_result(int a, const char *b) { (void)a; return b; }
-#else
+#if defined(STRERROR_R_CHAR_P)
 // GLIBC version of strerror_r
 static const char *strerror_result(const char *a, const char *b) { (void)b; return a; }
-#endif
-
-#else /* ! WITHOUT_C11_GENERIC */
+#elif defined(HAVE_STRERROR_R)
+// POSIX version of strerror_r
+static const char *strerror_result(int a, const char *b) { (void)a; return b; }
+#elif defined(HAVE_C__GENERIC)
 
 // what a trick!
 // http://stackoverflow.com/questions/479207/function-overloading-in-c
@@ -288,7 +284,9 @@ static const char *strerror_result_string(const char *a, const char *b) { (void)
     char *: strerror_result_string \
     )(a, b)
 
-#endif /* ! WITHOUT_C11_GENERIC */
+#else
+#error "cannot detect the format of function strerror_r()"
+#endif
 
 void error_int( const char *prefix, const char *file, const char *function, const unsigned long line, const char *fmt, ... )
 {

--- a/src/log.c
+++ b/src/log.c
@@ -266,6 +266,18 @@ void info_int( const char *file, const char *function, const unsigned long line,
 // ----------------------------------------------------------------------------
 // error log
 
+#ifdef WITHOUT_C11_GENERIC
+
+#ifdef STRERROR_R_POSIX
+// POSIX version of strerror_r
+static const char *strerror_result(int a, const char *b) { (void)a; return b; }
+#else
+// GLIBC version of strerror_r
+static const char *strerror_result(const char *a, const char *b) { (void)b; return a; }
+#endif
+
+#else /* ! WITHOUT_C11_GENERIC */
+
 // what a trick!
 // http://stackoverflow.com/questions/479207/function-overloading-in-c
 static const char *strerror_result_int(int a, const char *b) { (void)a; return b; }
@@ -275,6 +287,8 @@ static const char *strerror_result_string(const char *a, const char *b) { (void)
     int: strerror_result_int, \
     char *: strerror_result_string \
     )(a, b)
+
+#endif /* ! WITHOUT_C11_GENERIC */
 
 void error_int( const char *prefix, const char *file, const char *function, const unsigned long line, const char *fmt, ... )
 {

--- a/src/log.c
+++ b/src/log.c
@@ -266,6 +266,16 @@ void info_int( const char *file, const char *function, const unsigned long line,
 // ----------------------------------------------------------------------------
 // error log
 
+// what a trick!
+// http://stackoverflow.com/questions/479207/function-overloading-in-c
+static const char *strerror_result_int(int a, const char *b) { (void)a; return b; }
+static const char *strerror_result_string(const char *a, const char *b) { (void)b; return a; }
+
+#define strerror_result(a, b) _Generic((a), \
+    int: strerror_result_int, \
+    char *: strerror_result_string \
+    )(a, b)
+
 void error_int( const char *prefix, const char *file, const char *function, const unsigned long line, const char *fmt, ... )
 {
     va_list args;
@@ -283,12 +293,7 @@ void error_int( const char *prefix, const char *file, const char *function, cons
 
     if(errno) {
         char buf[1024];
-#if ((_POSIX_C_SOURCE >= 200112L) && !  _GNU_SOURCE)
-        strerror_r(errno, buf, 1023);
-        fprintf(stderr, " (errno %d, %s)\n", errno, buf);
-#else
-        fprintf(stderr, " (errno %d, %s)\n", errno, strerror_r(errno, buf, 1023));
-#endif
+        fprintf(stderr, " (errno %d, %s)\n", errno, strerror_result(strerror_r(errno, buf, 1023), buf));
         errno = 0;
     }
     else

--- a/src/log.c
+++ b/src/log.c
@@ -101,7 +101,7 @@ void reopen_all_log_files() {
 
 void open_all_log_files() {
     // disable stdin
-    open_log_file(STDIN_FILENO, &stdin, "/dev/null", NULL);
+    open_log_file(STDIN_FILENO, (FILE **)&stdin, "/dev/null", NULL);
 
     open_log_file(STDOUT_FILENO, (FILE **)&stdout, stdout_filename, &output_log_syslog);
     open_log_file(STDERR_FILENO, (FILE **)&stderr, stderr_filename, &error_log_syslog);
@@ -283,7 +283,12 @@ void error_int( const char *prefix, const char *file, const char *function, cons
 
     if(errno) {
         char buf[1024];
+#if ((_POSIX_C_SOURCE >= 200112L) && !  _GNU_SOURCE)
+        strerror_r(errno, buf, 1023);
+        fprintf(stderr, " (errno %d, %s)\n", errno, buf);
+#else
         fprintf(stderr, " (errno %d, %s)\n", errno, strerror_r(errno, buf, 1023));
+#endif
         errno = 0;
     }
     else

--- a/src/log.c
+++ b/src/log.c
@@ -90,22 +90,22 @@ int open_log_file(int fd, FILE **fp, const char *filename, int *enabled_syslog) 
 
 void reopen_all_log_files() {
     if(stdout_filename)
-        open_log_file(STDOUT_FILENO, &stdout, stdout_filename, &output_log_syslog);
+        open_log_file(STDOUT_FILENO, (FILE **)&stdout, stdout_filename, &output_log_syslog);
 
     if(stderr_filename)
-        open_log_file(STDERR_FILENO, &stderr, stderr_filename, &error_log_syslog);
+        open_log_file(STDERR_FILENO, (FILE **)&stderr, stderr_filename, &error_log_syslog);
 
     if(stdaccess_filename)
-        stdaccess_fd = open_log_file(stdaccess_fd, &stdaccess, stdaccess_filename, &access_log_syslog);
+        stdaccess_fd = open_log_file(stdaccess_fd, (FILE **)&stdaccess, stdaccess_filename, &access_log_syslog);
 }
 
 void open_all_log_files() {
     // disable stdin
     open_log_file(STDIN_FILENO, &stdin, "/dev/null", NULL);
 
-    open_log_file(STDOUT_FILENO, &stdout, stdout_filename, &output_log_syslog);
-    open_log_file(STDERR_FILENO, &stderr, stderr_filename, &error_log_syslog);
-    stdaccess_fd = open_log_file(stdaccess_fd, &stdaccess, stdaccess_filename, &access_log_syslog);
+    open_log_file(STDOUT_FILENO, (FILE **)&stdout, stdout_filename, &output_log_syslog);
+    open_log_file(STDERR_FILENO, (FILE **)&stderr, stderr_filename, &error_log_syslog);
+    stdaccess_fd = open_log_file(stdaccess_fd, (FILE **)&stdaccess, stdaccess_filename, &access_log_syslog);
 }
 
 // ----------------------------------------------------------------------------

--- a/src/web_client.c
+++ b/src/web_client.c
@@ -1903,8 +1903,18 @@ static inline int http_request_validate(struct web_client *w) {
 }
 
 void web_client_process(struct web_client *w) {
-    static uint32_t hash_api = 0, hash_netdata_conf = 0, hash_data = 0, hash_datasource = 0, hash_graph = 0,
-            hash_list = 0, hash_all_json = 0, hash_exit = 0, hash_debug = 0, hash_mirror = 0;
+    static uint32_t
+            hash_api = 0,
+            hash_netdata_conf = 0,
+            hash_data = 0,
+            hash_datasource = 0,
+            hash_graph = 0,
+            hash_list = 0,
+            hash_all_json = 0;
+
+#ifdef NETDATA_INTERNAL_CHECKS
+    static uint32_t hash_exit = 0, hash_debug = 0, hash_mirror = 0;
+#endif
 
     // start timing us
     gettimeofday(&w->tv_in, NULL);

--- a/src/web_client.c
+++ b/src/web_client.c
@@ -1927,9 +1927,11 @@ void web_client_process(struct web_client *w) {
         hash_graph = simple_hash(WEB_PATH_GRAPH);
         hash_list = simple_hash("list");
         hash_all_json = simple_hash("all.json");
+#ifdef NETDATA_INTERNAL_CHECKS
         hash_exit = simple_hash("exit");
         hash_debug = simple_hash("debug");
         hash_mirror = simple_hash("mirror");
+#endif
     }
 
     int code = 500;

